### ISSUE TITLE
Object spread produces "not assignable" error with exact property match

### ISSUE THREAD

**Issue by kyeotic** [microsoft/TypeScript#41557]
**TypeScript Version:**  4.0.5

**Search Terms:**
```
typescript object spread
typescript object spread not assignable
typescript object spread incompatible
typescript object literal valid object spread incompatible
```

**Expected behavior:** Object spread is no different than object literal with the exact same properties.

**Actual behavior:** Object spread produces `is not assignable to parameter` Error

```
Argument of type '{ responseBody: string; method: string; protocol: string; host: string; }' is not assignable to parameter of type 'DenockOptions'.
  Types of property 'method' are incompatible.
    Type 'string' is not assignable to type 'HTTPMethods'.
```

While object literal of the exact same properties does not produce an error.

**Related Issues:**
Similar, but not the same
[Object spread can lead to unknown properties added](https://github.com/microsoft/TypeScript/issues/41237)
[Spreading introduces undeclared fields in object of specific type](https://github.com/microsoft/TypeScript/issues/41237)

**Code**
```ts
type HTTPMethods = "GET" | "POST" | "PATCH" | "PUT" | "DELETE";

interface DenockOptions {
  method: HTTPMethods;
  protocol: "http" | "https";
  host: string;
  port?: number;
  path?: string;
  queryParams?: any;
  headers?: Array<{ header: string; value: string }>;
  /** This is the body that the request must contains to be intercepted*/
  requestBody?: any;
  /** This is the status code that will be returned on interception*/
  replyStatus?: number;
  /** This is the body that will be returned on interception*/
  responseBody: any;
  /** Represent the number of call you wants to intercept*/
  interception?: number;
}

function denock(options: DenockOptions): void {
    console.log(options)
    return
}

  const denockOptions = {
    method: "POST",
    protocol: "http",
    host: 'fake.com',
  };

  denock({
    ...denockOptions, // <--- fails
    responseBody: JSON.stringify({
      data: { },
    }),
  });

  denock({  // <--- works
    method: "POST",
    protocol: "http",
    host: 'fake.com',
    responseBody: JSON.stringify({
      data: { },
    }),
  });

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function denock(options) {
    console.log(options);
    return;
}
const denockOptions = {
    method: "POST",
    protocol: "http",
    host: 'fake.com',
};
denock(Object.assign(Object.assign({}, denockOptions), { responseBody: JSON.stringify({
        data: {},
    }) }));
denock({
    method: "POST",
    protocol: "http",
    host: 'fake.com',
    responseBody: JSON.stringify({
        data: {},
    }),
});

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/C4TwDgpgBAEgKnACgWQsAFgewCYGcoC8UARAOICicxUAPiYgPIDKVt9AgnAMIzV3GIAqq34ARcgBlK5YgG4AUPICWAO2AQATgDMAhgGNooiCsx6A1gzDAlmFfgDe8qFAC2aLNgBcsBCnc5cBWcwDUxgU0wAG29idGBgMD4SOITcOScoLFxgb2yNVQBzIKgwTA1gAH5vFQBXFwAjTWKwHQwqqDzC4oBHGs0QRB0NHRdcdp0VEGL0CB1sTTHvdg1hkAAee0zZ+Y1c4HyVIqgANx1Ivr2DgqgAXwA+YoB6ACpnqDh0JXwvqAxoepwIF+6FawOgGggvQg2VcNRhelswB0qnw4SgjSgqnUGgMVgg2GejwyEKh2QAQoDxpMnq93p9vqiZh0kcA4VAEfNgaCAO5KSKRdHgtA1DQqfFQWyYtSaXHWWyE4kQMCREBMFlw9q1BpNDIvN4fH4-P7owFc4BQXn8wVQCGs0XiyVYmVKuUqBXOCG4Up2CAU7AgbwTKa62kAJSVnuM5uNWsaGglWnZZwFIEwNQtE2AqMwUuxsvduedVhsKk1dTjChuii0NRUeldUHmJnMAApMMXbLhvEZmxYO3YAJTeY6YJTYKCOZzOBF2KIQAB0kUwBTb-dwA4yHuFovkVYyM5hTdMfdd+CIk6nbgwOBijBYxAANJuSqFwgjosl4okn1PMphst4ADkuhmAuCIuIBP63AoGRHq2F7OPOSFwSeJa4FBW5ep2vqAt4ABSTAMAAcvOnSHEoWggC2CFTtgrQ6N4mw3BhtwDlBNwDjBzgodRz5Xh4t7MFQLEhGEEQfrEX6Ps+WQ5FAwE6KB87gZBz6et6uA4f6+GESRZEFBRVE0dx9GMbcLEcexnHyEAA)

**Comment by RyanCavanaugh** [comment]
This is the intended behavior; the given declaration of `denockOptions` doesn't disallow mutations like

```
denockOptions.method = "upvote";
```

If you want to prevent that and thus let the code pass the checker, you can write
```ts
  const denockOptions = {
    method: "POST",
    protocol: "http",
    host: 'fake.com',
  } as const
    // ^^^^
```

**Comment by kyeotic** [comment]
Oh hey, I learned something new today 😄 

Is that in the handbook somewhere? I tried searching high and low for object spread quirks and didn't find this incredibly useful info.

**Comment by Validark** [comment]
This isn't an "object spread quirk". You declared an object without giving TS any constraints on what it should look like and the inferred type by default for `denockOptions.method` is `string`. Mouse-over the object name to see the inferred type. Or just look at the error you posted:

 ```
Types of property 'method' are incompatible.
    Type 'string' is not assignable to type 'HttpMethods'
```

^^ What do you think this is saying?

The reason adding `as const` "fixes it" is because it makes the object literal readonly and makes all the string types infer to the literal string in the definition. I.e it makes `denockOptions.method` of the type `"POST"` which is assignable to type 'HttpMethods'.

**Comment by kyeotic** [comment]
I thought it was an error because taking that same string and putting into the object literal version worked. 

Your explanation makes sense. However the fact that it works it an object literal but not as part of a spread object certainly makes it appear as if it’s a quirk of the method and not an effect of the type system. I had never considered that a string changes type when it’s container becomes immutable. Again, it makes sense, but when you move code from on untyped container to another untyped container and it stops working it looks like a quirk.
